### DISCUSSION TITLE
Import MDX components dynamically

### DISCUSSION THREAD

**Discussion by mingue** [vercel/next.js#12151] [Help]
Hi,

  I'm new to NextJs and although I find it extremely easy to use, you did an amazing work. I'm struggling with something a bit more advanced, maybe you can offer some guidance.

  I'm basically trying to create a website, with a blog section that allows me to load dynamically mdx files to have markdown with React components in it, but when I try to build the solution it fails saying that cannot find any file on the path, it looks like is trying to search for .tsx, .jsx files but not .mdx

  I've tested displaing MDX files by itself without importing them and they work fine, but I would like to load them dynamically so that I can keep the bundle light for the rest of the pages and only load MdxProvider and the rest of the JS needed to process the blog content when I'm displaying this content.

  This is what I've tried so far with no success:

```tsx
import React, { FunctionComponent } from "react";
import CodeBlock from "../../components/codeBlock";
import { MDXProvider } from "@mdx-js/react";
import dynamic from "next/dynamic";

const components = {
  code: CodeBlock,
};

const PostPage: FunctionComponent<{}> = ({}) => {
  const Post = dynamic(() =>
    import(
      /* webpackInclude: /\.mdx$/ */
      "../../content/post1"
    )
  );

  return (
    <MDXProvider components={components}>
      <Post />
    </MDXProvider>
  );
};

export default PostPage;
```

The idea is that the PostPage /blog/[slug].tsx component will receive a parameter by url and then try to load the relevant .mdx component.

This is the error I'm getting on build time:

```text
ModuleNotFoundError: Module not found: Error: Can't resolve '../../content/post1' in 'C:\Users\victo\Projects\next-blog\src\pages\blog'
    at C:\Users\victo\Projects\next-blog\src\node_modules\webpack\lib\Compilation.js:925:10
    at C:\Users\victo\Projects\next-blog\src\node_modules\webpack\lib\NormalModuleFactory.js:401:22
    at C:\Users\victo\Projects\next-blog\src\node_modules\webpack\lib\NormalModuleFactory.js:130:21
    at C:\Users\victo\Projects\next-blog\src\node_modules\webpack\lib\NormalModuleFactory.js:224:22
    at C:\Users\victo\Projects\next-blog\src\node_modules\neo-async\async.js:2830:7
    at C:\Users\victo\Projects\next-blog\src\node_modules\neo-async\async.js:6877:13
    at C:\Users\victo\Projects\next-blog\src\node_modules\webpack\lib\NormalModuleFactory.js:214:25
    at C:\Users\victo\Projects\next-blog\src\node_modules\enhanced-resolve\lib\Resolver.js:213:14
    at C:\Users\victo\Projects\next-blog\src\node_modules\enhanced-resolve\lib\Resolver.js:285:5
    at eval (eval at create (C:\Users\victo\Projects\next-blog\src\node_modules\tapable\lib\HookCodeFactory.js:33:10), <anonymous>:13:1)
resolve '../../content/post1' in 'C:\Users\victo\Projects\next-blog\src\pages\blog'
  using description file: C:\Users\victo\Projects\next-blog\src\package.json (relative path: ./pages/blog)
    Field 'browser' doesn't contain a valid alias configuration
    using description file: C:\Users\victo\Projects\next-blog\src\package.json (relative path: ./content/post1)
      no extension
        Field 'browser' doesn't contain a valid alias configuration
        C:\Users\victo\Projects\next-blog\src\content\post1 doesn't exist
      .tsx
        Field 'browser' doesn't contain a valid alias configuration
        C:\Users\victo\Projects\next-blog\src\content\post1.tsx doesn't exist
      .ts
        Field 'browser' doesn't contain a valid alias configuration
        C:\Users\victo\Projects\next-blog\src\content\post1.ts doesn't exist
      .mjs
        Field 'browser' doesn't contain a valid alias configuration
        C:\Users\victo\Projects\next-blog\src\content\post1.mjs doesn't exist
      .js
        Field 'browser' doesn't contain a valid alias configuration
        C:\Users\victo\Projects\next-blog\src\content\post1.js doesn't exist
      .jsx
        Field 'browser' doesn't contain a valid alias configuration
        C:\Users\victo\Projects\next-blog\src\content\post1.jsx doesn't exist
      .json
        Field 'browser' doesn't contain a valid alias configuration
        C:\Users\victo\Projects\next-blog\src\content\post1.json doesn't exist
      .wasm
        Field 'browser' doesn't contain a valid alias configuration
        C:\Users\victo\Projects\next-blog\src\content\post1.wasm doesn't exist
      as directory
        C:\Users\victo\Projects\next-blog\src\content\post1 doesn't exist
Creating an optimized production build .npm ERR! code ELIFECYCLE
```

Any help will be greatly appreciated :-)

**Comment by jemjam** [comment]
I came here to mention that I've been chasing after the same thing.

I have a folder of MDX files in a /posts subdirectory. I'd like for `pages/post/[...slug].tsx` to pass along the slug and load the relevant markdown file. This is a common pattern: reading markdown files from a directory. Unfortunately with mdx it's trickier if you intend to render live components, you need to be careful about how it's parsed or mounted. For instance: reading the contents with fs.readFile will only work if you plan to parse and include the entire mdx runtime in the pages, which is a several hundred kb payload, far from ideal to just parse out some content.

Another angle that I tried was using the next dynamic import. Dynamic imports work a treat when you have **completely static paths** to render. Things worked fine if I wanted to use `dynamic(() => import('posts/static/path')`, but the moment I tried to pass in any kind of dynamic parameter (like `import(someVarIJustSet)` things break down

While googling around I uncovered [this thread](https://github.com/faceyspacey/react-universal-component/issues/139#issuecomment-417842211) that mentions some of the issues inherent to webpack and dynamic imports. The linked comment in particular suggests building out a map of imports, to create a `pseudo-dynamic` import where at least each path is known ahead of time. With this approach I was able to sort of acheive success, but only by defining the map statically. The closest I've got to dynamic imports is a large hand rolled map of dynamic imports.

```
// posts/AllMap.ts
{
  'pageA': dynamic(() => import('posts/pageA.mdx')),
}
```

With that structure, you can pass the slug from the props and dynamically load the mdx. It works, but it's not ideal. I'll note that EVEN trying to have a simpler input map that I then loop over and build up (like `"pageA": "post/pageA"`  that then concats file extensions, or prefixes the dir) would not work. The moment the import() contains anything remotely dynamic or dynamically generated my mdx content would fail to load. (On that front I started to investigate whether somehow I could leverage webpack...)

Anyways... this might not help your case, but I've been all over the map and figured I'd share some of my attempts (in the hopes that one of us gets there in the end).

**Reply by mingue** [reply]
At the end opted for the simplest approach, although requires a tiny bit of boilerplate, seems to go well with the design of MDX itself, where you can load React components. This also give me the benefit that I was looking for, where it'll load only the JS needed to render the posts without affecting the other sections of web.

I opted for creating a wrapper component in React that will load all the dependencies needed for any MDX file, in my case looks something like:

```tsx
import React from "react";
import CodeBlock from "../codeBlock";
import { MDXProvider } from "@mdx-js/react";

const components = {
  code: CodeBlock,
};

export default ({ children }) => {
  return (
    <MDXProvider components={components}>
      <div className="section-wrapper">
        <div className="container-fluid section-title">{children}</div>
      </div>
    </MDXProvider>
  );
};
```

And then to use it, I only wrap my markdown with the wrapper like so:

```mdx
export const meta = {
  author: "Andrew Lisowski",
  authorLink: "https://github.intuit.com/alisowski",
  avatar: "https://avatars2.githubusercontent.com/u/1192452?s=400&v=4",
  publishDate: "2018-05-10T12:00:00Z",
  title: "First Post",
};

import MdxWrapper from "../../components/layout/mdx-wrapper.tsx";

<MdxWrapper>
  # Some content

  ```json
  {
    "prop": "value"
  }
  \```

  ```js
  import React from "react";
  import Example from "./place";

  export default () => <Example />;
  \```
</MdxWrapper>
```

Finally to load those mdx files, I just use the normal routing from Next.

Hope that helps!

**Reply by Vadorequest** [reply]
@jemjam Did you try using `__non_webpack_require__`?

https://stackoverflow.com/questions/46185302/non-webpack-require-is-not-defined#:~:text=__non_webpack_require__%20is%20a%20way%20to%20work%20around%20the,it%20actually%20is%20in%20Node.

Example of use: https://github.com/UnlyEd/next-right-now/blob/v2-mst-aptd-gcms-lcz-sty/src/utils/i18n/i18nextLocize.ts#L353

**Reply by helloworld** [reply]
Did anyone here ever come up with a better approach? I'm running into the same problem, and would love to hear about alternative solutions. Especially ones that take advantage of `getServerSideProps`

**Reply by treyhuffine** [reply]
@jemjam I've been doing something similar and realized that the `dynamic` didn't code split and loaded everything in the map even though they exist on different pages/slugs. Did you notice anything similar and find a way to overcome it?

**Comment by goldins** [comment]
Came across this page when trying to do something similar with exported `metadata` consts. Still unable to get that working, but for the content itself here's what I have. Hopefully it can be adapted and improved:

```
export default ({ fileNames }) => {
  const MDs = fileNames.map((name) => dynamic(() => import(`./content/${name}`)));
  return (
    <>
      {MDs.map((MD, index) => (
        <MD key={index} />
      ))}
    </>
  );
};
```

I'm setting `fileNames` in `getStaticProps`. Essentially:
```
export const getStaticProps = async () => ({
  props: { fileNames: fs.readdirSync(postsDirectory) }
})
```

`./content` is the name of my directory relative to this index page.

Edit: a few hours later, I was able to import `metadata` like this:
```
const metas = fileNames.map((name) => require(`./content/${name}`).metadata);
```

I'm not sure if I can get the content and metadata at once.

**Reply by daveteu** [reply]
you can give nextjs-mdx-extended / next-mdx-enhanced a try, I believe it can support frontmatter

**Comment by Vadorequest** [comment]
I'm trying to do something similar at https://github.com/vercel/next.js/discussions/13843, managed to do it with static paths, want to make it work with content from a CMS now.

Related:
- https://github.com/vercel/next.js/discussions/13901
- https://github.com/mdx-js/mdx/issues/126
- The official MDX doc seem to use a virtual file (need to dig in, may be the solution): https://mdxjs.com/advanced/api#compiler

**Comment by CallumJHays** [comment]
I managed to get this working with the code below. The tricky part was the line:
```typescript
const meta: FrontMatter = (await import(`pages/blog/${filename}.mdx`)).frontMatter;
```
It didn't work when I had:
```typescript
const meta: FrontMatter = (await import(path)).frontMatter;
```
Even though they both resolved to the same string.

Full code:
```typescript
import fg from "fast-glob";
import { IEntry } from "fast-glob/out/types/entries";
import { GetStaticProps } from "next";

export default function BlogIndexPage({ blogPosts }) { ... } // etc

export const getStaticProps: GetStaticProps<Props> = async () => {

  const blogPosts = await Promise.all(
    (await fg("pages/blog/*.mdx", { stats: true })).map(
      
      async (file: IEntry) => {
        const { path, mtimeMs, birthtimeMs } = file;
        const filename = path.match(/blog\/(.*)+\.mdx/)[1];

        // this line is very brittle. NextJS doin' some funky magic
        const meta: FrontMatter = (await import(`pages/blog/${filename}.mdx`)).frontMatter;

        return {
          url: `/blog/${filename}`,
          createdMs: birthtimeMs,
          editedMs: mtimeMs,
          ...meta,
        };
      }
    )
  );

  return {
    props: {
      blogPosts,
    },
  };
};
```
(using next@10.0.4)

**Reply by michrome** [reply]
Rather than Next.js doin' some funky magic 🕺, I _think_ the reason the **directory** we're importing from needs to be a hard-coded string to avoid webpack looking at all directories. That's why we can use:

```javascript
await import(`pages/blog/${fileName}`)
```

but not

```javascript
await import(directoryAndFileName)
```

or

```javascript
await import(join(process.cwd(), "pages", "blog", fileName))
```

**Comment by AnvarNazar** [comment]
I have successfully managed to import mdx files as components dynamically using the [dynamic import](https://nextjs.org/docs/advanced-features/dynamic-import) feature of Next.js

/pages/posts/[id].js
```
import Date from "@/components/date";
import Layout from "@/components/layout";
import { getAllPostIds, getPostData } from "@/lib/posts";
import dynamic from "next/dynamic";
import Head from "next/head";

export async function getStaticProps({ params }) {
    const postData = await getPostData(params.id);
    return {
        props: {
            postData,
        },
    };
}

export async function getStaticPaths() {
    const paths = getAllPostIds();
    return {
        paths,
        fallback: false,
    };
}

export default function Post({ postData }) {
    const Mdx = dynamic(import(`@/posts/${postData.fileName}`), { ssr: false });
    return (
        <Layout>
            <Head>
                <title>{postData.title}</title>
            </Head>
            {postData.title}
            <br />
            {postData.id}
            <br />
            <Date dateString={postData.date} />
            <Mdx />
        </Layout>
    );
}
```
/lib/posts.js
```
export async function getPostData(id) {
	const fileName = `${id}.mdx`;
    const fullPath = path.join(postsDirectory, fileName);
    const fileContents = fs.readFileSync(fullPath, "utf8");

    // Use gray-matter to parse the post metadata section
    const matterResult = matter(fileContents);

    // Combine the data with the id and contentHtml
    return {
		fileName,
        id,
        ...matterResult.data,
    };
}

export function getAllPostIds() {
    const fileNames = fs.readdirSync(postsDirectory);

    // Returns an array that looks like this:
    // [
    //   {
    //     params: {
    //       id: 'ssg-ssr'
    //     }
    //   },
    //   {
    //     params: {
    //       id: 'pre-rendering'
    //     }
    //   }
    // ]
    return fileNames.map((fileName) => {
        return {
            params: {
                id: fileName.replace(/\.mdx$/, ""),
            },
        };
    });
}
```

> I am not using SSR here. But it should work on it. Refer the dynamic import documentation.

**Reply by 4ndv** [reply]
For anyone wandering around trying to understand why ssr is not working with this (after removing ssr: false): it would not, https://github.com/vercel/next.js/discussions/32242
